core/preconditioner/block_jacobi.hpp

```diff
@@ -192,17 +276,59 @@ class BasicBlockJacobi : public EnableLinOp<ConcreteBlockJacobi> {
               exec, transpose(system_matrix->get_size())),
           num_blocks_(block_pointers.get_num_elems() - 1),
           max_block_size_(max_block_size),
+          storage_scheme_{compute_storage_scheme(max_block_size)},
           block_pointers_(block_pointers),
-          blocks_(exec, this->get_size()[1] * max_block_size)
+          blocks_(exec, storage_scheme_.compute_storage_space(
+                            block_pointers.get_num_elems() - 1))
     {
         block_pointers_.set_executor(this->get_executor());
     }
 
-protected:
     size_type num_blocks_{};
     uint32 max_block_size_{};
+    block_interleaved_storage_scheme storage_scheme_;
     Array<index_type> block_pointers_;
     Array<value_type> blocks_;
+
+    /**
+     * Stride between two columns of a block (as number of elements).
+     *
+     * Should be a multiple of cache line size for best performance.
```

Comment:
The cache line size varies across machines and different hardware right? Would you not like to have this as a parameter then? Or have I misunderstood something ?

Reply:
I don't know of tools for that, there surely is some, but for the CPU you can at least find the information here on Linux:
```
/sys/devices/system/cpu/cpu0/cache/index0/coherency_line_size
```

Mine says 64 bytes for example. We could either get this information statically through CMake (but you have to compile on the final system) or use some executable/functions to get the information dynamically.

There is also this tool for the CPU which has Linux, MacOS and Windows compatibility.
https://github.com/NickStrupat/CacheLineSize